crates/ruma-common/src/canonical_json.rs

```diff
@@ -0,0 +1,367 @@
+//! Canonical JSON types and related functions.
+
+use std::{fmt, mem};
+
+use serde::Serialize;
+use serde_json::Value as JsonValue;
+
+mod value;
+
+use crate::RoomVersionId;
+
+pub use self::value::{CanonicalJsonObject, CanonicalJsonValue};
+
+/// The set of possible errors when serializing to canonical JSON.
+#[cfg(feature = "canonical-json")]
+#[derive(Debug)]
+#[allow(clippy::exhaustive_enums)]
+pub enum CanonicalJsonError {
+    /// The numeric value failed conversion to js_int::Int.
+    IntConvert,
+
+    /// An error occurred while serializing/deserializing.
+    SerDe(serde_json::Error),
+}
+
+impl fmt::Display for CanonicalJsonError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CanonicalJsonError::IntConvert => {
+                f.write_str("number found is not a valid `js_int::Int`")
+            }
+            CanonicalJsonError::SerDe(err) => write!(f, "serde Error: {err}"),
+        }
+    }
+}
+
+impl std::error::Error for CanonicalJsonError {}
+
+/// Errors that can happen in redaction.
+#[cfg(feature = "canonical-json")]
+#[derive(Debug)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum RedactionError {
+    /// The field `field` is not of the correct type `of_type` ([`JsonType`]).
+    #[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+    NotOfType {
+        /// The field name.
+        field: String,
+        /// The expected JSON type.
+        of_type: JsonType,
+    },
+
+    /// The given required field is missing from a JSON object.
+    JsonFieldMissingFromObject(String),
+}
+
+impl fmt::Display for RedactionError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RedactionError::NotOfType { field, of_type } => {
+                write!(f, "Value in {field:?} must be a JSON {of_type:?}")
+            }
+            RedactionError::JsonFieldMissingFromObject(field) => {
+                write!(f, "JSON object must contain the field {field:?}")
+            }
+        }
+    }
+}
+
+impl std::error::Error for RedactionError {}
+
+impl RedactionError {
+    fn not_of_type(target: &str, of_type: JsonType) -> Self {
```

Comment:
Nit: could probably be `AsRef<str>`, same below

Reply:
It's just a private helper function 🤷🏼

Reply:
Ah it's private, didn't see that, then this is fine 👍